root-cert-stats/background.js

```diff
@@ -0,0 +1,37 @@
+"use strict";
+
+var rootCertStats = {};
+
+/*
+On an onHeadersReceived event, if there was a successful TLS connection
+established, fetch the root cert and look at its subject.
+
+If we haven't seen this subject before, add it. If we have, increment its stats.
+*/
+async function logRootCert(details) {
+  try {
+    let securityInfo = await browser.webRequest.getSecurityInfo(
+      details.requestId,
+      {"certificateChain": true}
+    );
+    if (securityInfo.state == "secure" || securityInfo.state == "weak") {
+      let rootName = securityInfo.certificates[securityInfo.certificates.length - 1].subject;
```

Comment:
How about also keep track of when `securityInfo.isUntrusted  === true`, so that we can make those entries "red colored" once rendered in the popup?

e.g. something like:

```
var rootCertStats = {};
var untrustedCerts = new Set();
...
      let rootName = securityInfo.certificates[securityInfo.certificates.length - 1].issuer;
      ...
      if (securityInfo.isUntrusted) {
        untrustedCerts.add(rootName);
      }
      ...
``` 

and then use that set to highlight the untrusted entries in the popup:

```
    ...
    let entryTR = document.createElement("tr");
    let entryName =  document.createElement("td");
    let entryValue =  document.createElement("td");

    if (backgroundPage.untrustedCerts.has(entry)) {
      entryTR.classList.add("untrusted");
    }
    ...
```

If we add this "feature" to the example, we may also add somewhere in the README.md file a suggestion to navigate a tab on the example pages provided by badssl.com (in particular: https://self-signed.badssl.com/ and https://untrusted-root.badssl.com/), to see the untrusted entries highlighted in the table rendered in the extension popup .

Reply:
One practical difficulty with this is that (I think) at present we can't get security information if the handshake fails (see https://bugzilla.mozilla.org/show_bug.cgi?id=1322748#c116 and https://bugzilla.mozilla.org/show_bug.cgi?id=1474657).

But even if we could, I'm not sure if this is a good plan. Currently this shows which trust roots are most used, assuming that if the connection worked then the last cert in `certificates[certificates.length-1]` is a trust root.

But if a chain to a trust root couldn't be constructed, then there isn't a trust root to show. I don't actually know what `certificates[certificates.length-1]` would contain in this case. Perhaps the last cert in the list sent from the server? It seems confusing to list that alongside the trust roots.

What I mean is: if a trust root wasn't found, then `certificates[certificates.length-1]` isn't a potential trust root - it's an intermediate or server cert whose issuer is untrusted.

I think, unless I misunderstand things.

Reply:
yeah, the user has to add an exception and reload the page to allow the extension to see the security information in case of failed handshakes (and not all the failed handshakes can be bypassed with an exception actually).

I think that you are right, looking at the code behind the `isUntrusted` property, securityInfo.isUntrusted doesn't seem to be enough to be sure if the issuse was because of an unknown/invalid CA or because of one of the other issues that are going to set it to true (e.g. a missing issuer, inadequate key size etc.)

I'm thinking that it may be more reasonable to do not collect the root certificate from a securityInfo that has `isUntrusted === true`.

Reply:
Yes, that makes sense. I had assumed that if `isUntrusted === true` then `state` would be "broken", but if you visit https://untrusted-root.badssl.com/ and add the security exception, then `state` is apparently marked as "secure".

(Which seems a bit weird, now I think about it. Would it be better to have a new value for "state" like "security-exception"?)

Anyway, I've added a check so we won't collect the root if `isUntrusted === true`.